realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt

```diff
@@ -475,17 +473,90 @@ class CoroutineTests {
-        val otherMainScope = CoroutineScope(Dispatchers.Main)
-        otherMainScope.launch {
-            // Wait for 50 ms and cancel scope so that not all planned 10 elements are inserted
+        CoroutineScope(Dispatchers.Main).launch {
+            // Wait for 50 ms and cancel job so that not all planned 10 elements are inserted
             delay(50)
-            mainScope.cancel("Cancelling")
+            job.cancelAndJoin()
+
+            assertNotEquals(upperBound.toLong(), realmInstance!!.where<SimpleClass>().count())
+
+            realmInstance!!.close()
+
+            countDownLatch.countDown()
+            this.cancel()
+        }
+
+        TestHelper.awaitOrFail(countDownLatch)
+    }
+
+    @Test
+    fun executeTransactionAwait_cancelCoroutineWithHeavyCooperativeTransaction() {
+        val upperBound = 10000
+        var realmInstance: Realm? = null
+
+        val job = CoroutineScope(Dispatchers.Main).launch {
+            realmInstance = Realm.getInstance(configuration)
+
+            realmInstance!!.executeTransactionAwait { transactionRealm ->
+                for (i in 1..upperBound) {
+                    // The coroutine itself will not cancel the transaction, but we can make it cooperative ourselves
+                    if (isActive) {
+                        val simpleObject = SimpleClass().apply { name = "simpleName $i" }
+                        transactionRealm.insert(simpleObject)
+                    }
+                }
+            }
+        }
 
+        val countDownLatch = CountDownLatch(1)
+        CoroutineScope(Dispatchers.Main).launch {
+            // Wait for 50 ms and cancel job
+            delay(10)
+            job.cancelAndJoin()
+
+            // The coroutine won't finish until the transaction is completely done
+            // but not all elements will have been inserted since the transaction is cooperative
             assertNotEquals(upperBound.toLong(), realmInstance!!.where<SimpleClass>().count())
 
             realmInstance!!.close()
+
             countDownLatch.countDown()
-            otherMainScope.cancel()
+            this.cancel()
+        }
+
+        TestHelper.awaitOrFail(countDownLatch)
+    }
+
+    @Test
+    fun executeTransactionAwait_cancelCoroutineWithHeavyNonCooperativeTransaction() {
+        val upperBound = 10000
+        var realmInstance: Realm? = null
+
+        val job = CoroutineScope(Dispatchers.Main).launch {
+            realmInstance = Realm.getInstance(configuration)
+
+            realmInstance!!.executeTransactionAwait { transactionRealm ->
+                for (i in 1..upperBound) {
+                    val simpleObject = SimpleClass().apply { name = "simpleName $i" }
+                    transactionRealm.insert(simpleObject)
+                }
+            }
+        }
+
+        val countDownLatch = CountDownLatch(1)
+        CoroutineScope(Dispatchers.Main).launch {
+            // Wait for 50 ms and cancel job
+            delay(50)
```

Comment:
This delay feels a bit dangerous. 50ms is a long time and the first job might already have completed here?

Reply:
True. I wrote 50ms because I had another delay of 10ms in the loop of transactions in the original test, which was enough. I will lower it.